docs/source/guides/schema-design.md

```diff
@@ -0,0 +1,290 @@
+
+```graphql
+mutation updateMyUser {
+  updateUserAge(id: 1, age: 25){
+    id
+    age
+    name
+  }
+}
+```
+
+With a response looking something like:
+
+```json
+{
+  "data": {
+    "updateUserAge": {
+      "id": "1",
+      "age": "25",
+      "name": "jane doe"
+    }
+  }
+}
+```
+
+The first thing to note is that it’s most common to return the thing you’re updating from a mutation. In our example, we were updating a `User` record, so we returned that updated user. There’s nothing _enforcing_ this practice, but it’s highly recommended, because it’s often needed to have an updated user to let the clients update any local cache with a new instance of that `User`. For this same reason, it’s useful design mutations to update only one entity. If you wanted to update two unrelated entities, it’s recommended to in separate mutations. For more details on how to handle errors and warnings in mutations, see the [mutation responses](#mutation-responses) section below.
+
+But what if we wanted to update more than just a single or couple attributes on a user? Passing each thing we need as a single argument would get tedious. Especially if multiple mutations used similar fields. For this, we can use input types, which are explained in the next section.
+
+<h3 id="mutation-input-types">Input types</h3>
+
+Input types are a special type in GraphQL which are defined as arguments to queries and, more commonly, mutations.  They can be thought of as object types for arguments, in addition to the other scalar types.  Input types are especially useful when multiple mutations require similar information; for example, when creating a user and updating a user require the same fields, like `age` and `name`.
+
+Input types are used like any other type and defining them is similar to a typical object type definitions, but with the `input` keyword rather than `type`.
+
+Here is an example of two mutations that operate on a `User`, _without_ using input types:
+
+```
+type Mutation {
+  createUser(name: String, age: Int, address: String, phone: String): User
+  updateUser(id: ID!, name: String, age: Int, address: String, phone: String): User
+}
+```
+
+To avoid the repetition of argument fields, this can be refactored to use an input type, as follows:
+
+```
+type Mutation {
+  createUser(user: UserInput): User
+  updateUser(id: ID!, user: UserInput): User
+}
+
+input UserInput {
+  name: String
+  age: Int
+  address: String
+  phone: String
+}
+```
+
+<h3 id="mutation-responses">Responses</h3>
+
+Mutations have a higher chance of causing errors than queries since they are modifying data.  A common way to handle errors during a mutation is to simply `throw` an error.  While that's fine, throwing an error in the resolver will return an error to the caller and prevent a partial response, which could be useful in the event of a partial update.  Consider the following mutation example, which tries to update a user's `name` and `age`:
```

Comment:
This is definitely a good reason to have a response object type and not just returning the object directly, but there are two more equally compelling reasons:

1. A mutation might actually end up changing multiple things, like for example favoriting a post might update that post _and_ the list of a user's favorite posts, meaning it should probably return both the post and the user
2. A mutation return value might acquire more fields over time, and if you just return the User directly, you won't have an opportunity to add any more fields there later.

I think it would be good to mention these up front, and say that only the simplest mutations should return a scalar or object type directly; probably most mutations should return a special object type tailored specifically for that mutation.